### DISCUSSION TITLE
how to decode a str message : "01030200057847" ?

### DISCUSSION THREAD

**Discussion by patfrench** [pymodbus-dev/pymodbus#1899] [General]
Hi

I use pymodbus with the WEBSERIAL API, I build my messages with buidPacket() and it works very well.

My response (javascript) is in this format: "01 03 02 00 05 78 47".
I can't find in the documentation how to decode this message (str format). 

I'd like to read the data (0x05)

Can anyone help me?

thanks

**Comment by janiversen** [comment]
pymodbus do not have a web serial API, you are using a different project.

Seems you get the raw modbus message, you can look at our documentation to find the structure....pymodbus would return the user data.

**Comment by patfrench** [comment]
I use pymodbus in my web browser with pyscript.

``` 
<html>
  <head>
    <meta charset="utf-8" />
      <script defer src="https://pyscript.net/releases/2023.05.1/pyscript.js"></script>
        <link rel="stylesheet" href="https://pyscript.net/releases/2023.05.1/pyscript.css">
  </head>
  <body>
    <py-config>
        packages = ['pymodbus']
    </py-config>
    <py-script src="./main.py"></py-script>

    <py-repl id="my-repl" auto-generate="true"> </py-repl>
  </body>
</html>
```
pymodbus is installed in my browser but I use WEB SERIAL API instead of pyserial 

![image](https://github.com/pymodbus-dev/pymodbus/assets/44723412/0544b9f6-22da-49f1-9c36-020ea5b9d7a1)

test my demo : [https://titeppat.pyscriptapps.com/twilight-hill/latest/](https://titeppat.pyscriptapps.com/twilight-hill/latest/)

**Reply by janiversen** [reply]
That is not something we support, but it may work.

Seems you are using REPL, please read the documentation of the possible commands. In general REPL would not return the raw modbus message but the data, so something is not working in your setup.

Normally you would not use REPL, but make direct call like read_coils etc, please see the documentation and examples.

I am quite surprised that you have been able to change pyserial without code changes, we depend quite a lot on the specific way pyserial works.

If/when you get it to work, it would a nice addon for pymodbus, as you might have noted we have added a web based simulator and in the midterm it is intended to replace REPL.

**Reply by janiversen** [reply]
It might be easier for you first to play around with REPL or make a small python app, that does what you and then integrate that into your web solution.

**Comment by patfrench** [comment]
I'm sorry, but I can't explain my problem (i'm french :-) )

I'll update my demo tomorrow. 

I haven't modified pyserial, I use pymodbus functions to generate my modbus frames.

**Step 1. build my message**
```
message = modbus_register.ReadHoldingRegistersRequest(address=1276, count=1, slave=1)
framer = ModbusRtuFramer(None)
raw_packet = framer.buildPacket(message)
packet = c.encode(raw_packet, "hex_codec").decode("utf-8")
print(f"{message.__class__.__name__:44} = {packet}")
```

**step2. function javascript "serialwrite" send my packet (010304fc0001450a) on my serial port** 

````
function javascript (packet){
     serialweb.write(packet)
}
````

**step3. I receive the response in str format (010302000ff840)**

> I search in the documentation how to decode this answer  
> add = 1
> code read = 03
> nbr_byte = 2
> data = 000F = 15
> crc = F840

I'll update my demo tomorrow but I'll need a modbus rtu device and a 485 converter.

Thank you for your support

**Reply by janiversen** [reply]
If you use payload to generate the modbus frames and sends the frame directly you do not really use pymodbus.

pymodbus is intented to be used in a way where you do high level calls and receive responses.

Have a look at our examples, one of them decodes frames....but be aware that all the modbus logic implemented (like error handling, retry, timing), is something you need to implement yourself, when you work with frames

**Reply by janiversen** [reply]
Please be aware that your steps probably only will work once, because the methods you call set internal variables like e.g. timer values both monitor the transport write as well as handling the response, so building a second request will most likely cause problems.

Anyhow I must emphasize that you use the library in a way that is unsupported,
meaning we cannot help you.

**Reply by patfrench** [reply]
Actually my idea was to use modbus_mapper (one of your examples) but intercept the frame before sending pyserial.

Pyserial doesn't exist for web browsers.

I don't know the mechanics of pymodbus, I wanted to rebuild a new frame with the next address and continue sending and receiving in javascript.

**Reply by janiversen** [reply]
Good luck, that is quite a task you have ahead of you.

**Reply by patfrench** [reply]
Hi,

I update my demo [https://titeppat.pyscriptapps.com/twilight-hill/latest/](https://titeppat.pyscriptapps.com/twilight-hill/latest/)
click on "view code" at bottom right

works well with a **modbus rtu device at 9600 bauds**

I have a small problem with the first send but it's just a demo.
I realize I've got the wrong idea, but without pyserial I don't know how else to do it.

Thanks anyway

![image](https://github.com/pymodbus-dev/pymodbus/assets/44723412/ae710e1f-8cbb-42dd-874e-9acc08241cf3)

**Reply by janiversen** [reply]
Well there are a number of better ways, like e.g. make your own transport, but if it's work for you, then why bother.

But please try some error situations, that is the more interesting part.
